### ISSUE TITLE
Panic querying entity created in previous startup system

### ISSUE THREAD

**Issue by Cyborus04** [bevyengine/bevy#10869]
## Bevy version

v0.12.1 (crates.io)

## What you did

Querying for an entity created earlier in startup fails, even when using `before` and `after`.

The following code example

```rust
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, create_entity.before(search_for_entity))
        .add_systems(Startup, search_for_entity.after(create_entity))
        .run();
}

#[derive(Component)]
struct MyComponent;

fn create_entity(mut commands: Commands) {
    commands.spawn(MyComponent);
    println!("entity created");
}

fn search_for_entity(query: Query<&MyComponent>) {
    println!("querying entity");
    let _ = query.single();
}
```

outputs 

```
entity created
querying entity
thread 'Compute Task Pool (4)' panicked at src\main.rs:21:19:
called `Result::unwrap()` on an `Err` value: NoEntities("bevy_ecs::query::state::QueryState<&example::MyComponent>")
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Encountered a panic in system `example::search_for_entity`!
Encountered a panic in system `bevy_app::main_schedule::Main::run_main`!
```
Using `Query::iter` also fails to find the entity, it is not special to `Query::single`.

## What went wrong

I would have expected querying the entity would have succeeded, since it was created in a system known to have run before, but it just panicked with `NoEntities`

## Additional information

My guess is that the entity list isn't being updated until the end of the frame, but my knowledge of Bevy's internals is quite limited, so that could easily be incorrect.

**Comment by ItsDoot** [comment]
The issue is that all `Command`s are deferred until the end of the schedule (`Startup` in this case), OR until an inserted `apply_deferred` system runs.

In Bevy `0.12`, you can fix this pretty succinctly:
```rust
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        // .chain() runs the systems in the order they are supplied:
        .add_systems(Startup, (create_entity, apply_deferred, search_for_entity).chain())
        .run();
}
```
This will ensure the commands have been applied before the next system that needs them

---

Once #9822 is merged, you can turn that into:
```diff
fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
-       .add_systems(Startup, (create_entity, apply_deferred, search_for_entity).chain())
+       .add_systems(Startup, (create_entity, search_for_entity).chain())
        .run();
}
```
Because it will automatically insert `apply_deferred` when ordering constraints are defined (`chain` is the ordering constraint in this case, but it would also work with `after`/`before`.

**Comment by alice-i-cecile** [comment]
This is a common misconception. Will be fixed by https://github.com/bevyengine/bevy/pull/9822

**Comment by Cyborus04** [comment]
Thank you!
